core/ledger/kvledger/snapshot_mgmt.go

```diff
@@ -63,6 +90,154 @@ func (l *kvLedger) DeleteSnapshot(height uint64) error {
+			toStartCommitHeight = e.height
+			if snapshotInProgress {
+				logger.Infow("commit waiting on snapshot to be generated", "snapshotHeight", committedHeight)
+				continue
+			}
+			// no in-progress snapshot, let commit proceed
+			commitProceed <- struct{}{}
+
+		case commitDone:
+			committedHeight = e.height
+			if committedHeight != l.snapshotMgr.snapshotRequestBookkeeper.smallestRequestHeight {
+				continue
+			}
+			snapshotInProgress = true
+			requestedHeight := committedHeight
+			go func() {
+				if err := l.generateSnapshot(); err != nil {
+					logger.Errorw("Failed to generate snapshot", "height", requestedHeight, "error", err)
+				}
+				events <- &event{snapshotDone, requestedHeight}
+			}()
+
+		case snapshotDone:
+			requestedHeight := e.height
+			logger.Debugw("snapshot is generated", "height", requestedHeight, "toStartCommitHeight", toStartCommitHeight, "committedHeight", committedHeight)
+			if err := l.snapshotMgr.snapshotRequestBookkeeper.delete(requestedHeight); err != nil {
+				logger.Errorw("Failed to delete snapshot request, the pending snapshot requests (if any) may not be processed", "height", requestedHeight, "error", err)
+			}
+			if toStartCommitHeight != committedHeight {
+				// there is an in-progress commit, write to commitProceed channel to unblock commit
+				commitProceed <- struct{}{}
+			}
+			snapshotInProgress = false
+
+		case requestAdd:
+			requestedHeight := e.height
+			if requestedHeight == 0 {
+				requestedHeight = toStartCommitHeight
+			}
+			if requestedHeight < toStartCommitHeight {
+				requestResponses <- &requestResponse{errors.Errorf("requested snapshot height %d cannot be less than the current block height %d", requestedHeight, toStartCommitHeight)}
+				continue
+			}
+			if requestedHeight == committedHeight {
+				// this is a corner case where no block has been committed since last snapshot was generated.
+				exists, err := l.snapshotExists(requestedHeight)
+				if err != nil {
+					requestResponses <- &requestResponse{err}
+					continue
+				}
+				if exists {
+					requestResponses <- &requestResponse{errors.Errorf("snapshot already generated for block height %d", requestedHeight)}
+					continue
+				}
+			}
+			if err := l.snapshotMgr.snapshotRequestBookkeeper.add(requestedHeight); err != nil {
+				requestResponses <- &requestResponse{err}
+				continue
+			}
+			if toStartCommitHeight == committedHeight && requestedHeight == committedHeight {
+				snapshotInProgress = true
+				go func() {
+					if err := l.generateSnapshot(); err != nil {
+						logger.Errorw("Failed to generate snapshot", "height", requestedHeight, "error", err)
+					}
+					events <- &event{snapshotDone, requestedHeight}
+				}()
+				requestResponses <- &requestResponse{}
+				continue
+			}
+			// there is an in-progress commit or the requested height is higher than block height,
+			// we will evaluate the request at commitDone event
```

Comment:
if you remove the above lines, this comment can go off as well. With the revised if condition, this is clearer now anyways.